src/AI/JakeMan.java

```diff
@@ -785,16 +867,188 @@ private Map<XYCoord, UnitModel> queueUnitProductionActions(GameMap gameMap)
+      {
+        for( var fts : potentialMdBuilds )
+          fts.trackUnit(threat);
+        final UnitModel um = threat.model;
+        if( !allCounterContestants.contains(um) )
+          continue;
+        int oldVal = mapToFill.getOrDefault(um, 0);
+        int newVal = oldVal + threat.getHealth();
+        mapToFill.put(um, newVal);
+      }
+    }
+
+    boolean shouldSave = true;
+    for( var fts : potentialMdBuilds )
+    {
+      if( !fts.shouldBuildMd() )
+        continue;
+      MapLocation loc = gameMap.getLocation(fts.coord);
+      Commander buyer = loc.getOwner();
+      final XYCoord coord = loc.getCoordinates();
+      final int cost = buyer.getBuyCost(mdTank, coord);
+
+      int marginalCost = cost;
+      if( builds.containsKey(coord) )
+        marginalCost -= buyer.getBuyCost(builds.get(coord), coord);
+
+      if( marginalCost <= budget )
+      {
+        builds.put(coord, mdTank);
+        budget -= marginalCost;
+        shouldSave = false; // If we've already built some counter units, don't worry about saving for more
+        log(String.format("Building MD to protect %s", fts.coord));
+      }
+    }
+
+    boolean counterNeeded = false;
+    // If I need to do a counterbuild, either:
+    //   add it to the list before the normal builds
+    //   reduce my budget this turn to save up
+    for( var threatType : meanHealth.keySet() )
+    {
+      if( !unitTypeToCounterMap.containsKey(threatType) )
+        continue; // This isn't on our list of things to counter (e.g. AA)
+      int remainingHealth = meanHealth.get(threatType);
+      ArrayList<CounterRatio> counters = unitTypeToCounterMap.get(threatType);
+
+      // Chip down remainingHealth based on unit totals.
+      for( var ratio : counters )
+      {
+        if( !niceHealth.containsKey(ratio.counter) )
+          continue;
+        int counterHealth = niceHealth.get(ratio.counter);
+        if( counterTypeNegationMap.containsKey(threatType) )
+        {
+          var negationRatios = counterTypeNegationMap.get(threatType).getOrDefault(ratio.counter, new ArrayList<>());
+          for( var ccRatio : negationRatios )
+          {
+            if( !meanHealth.containsKey(ccRatio.counter) )
+              continue;
+            int ccHealth = meanHealth.get(ccRatio.counter);
+            int ccPower  = ccHealth * ccRatio.power / UnitModel.MAXIMUM_HEALTH;
+            counterHealth -= ccPower;
+          }
+        }
+        int counterPower  = counterHealth * ratio.power / UnitModel.MAXIMUM_HEALTH;
+        remainingHealth  -= counterPower;
+        if( ratio.roundTargetPercentUp )
+        {
+          int roundable = remainingHealth % UnitModel.MAXIMUM_HEALTH;
+          if( roundable >= CounterRatio.ROUND_UP_HEALTH )
+            remainingHealth += UnitModel.MAXIMUM_HEALTH - roundable;
+        }
+      }
+      if( remainingHealth < UnitModel.MAXIMUM_HEALTH )
+        continue; // We consider it fully-countered
```

Comment:
If a unit type has less than one unit's worth of cumulative HP left, Jakeman ignores it?

Reply:
Yes, because it enables him to use different rounding modes for different units.
IIRC fighter and bopter (as counter units) are the relevant examples.